verification/cbmc/proofs/aws_cryptosdk_hdr_write/aws_cryptosdk_hdr_write_harness.c

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/cryptosdk/header.h>
+#include <aws/cryptosdk/private/header.h>
+#include <make_common_data_structures.h>
+#include <proof_helpers/make_common_data_structures.h>
+#include <proof_helpers/proof_allocators.h>
+#include <proof_helpers/utils.h>
+
+/**
+ * A generator function as described in the comment in aws_cryptosdk_hash_elems_array_init_stub.c.
+ * Also see line 33 of the Makefile.
+ */
+void array_list_item_generator(struct aws_array_list *elems) {
+    assert(elems->item_size == sizeof(struct aws_hash_element));
+    for (size_t index = 0; index < elems->length; ++index) {
+        struct aws_hash_element *val = (struct aws_hash_element *)((uint8_t *)elems->data + (elems->item_size * index));
+        /* Due to the checks in aws_cryptosdk_enc_ctx_size, no string can have a length > UINT16_MAX. */
+        struct aws_string *key = ensure_string_is_allocated_nondet_length();
+        __CPROVER_assume(aws_string_is_valid(key));
+        /* Due to the cast to uint16, the entire size of the enc_ctx must be less than < UINT16_MAX
+         * This is a simple way to ensure this without a call to enc_ctx_size. */
+        __CPROVER_assume(key->len <= UINT8_MAX);
+        val->key                 = key;
```

Comment:
Maybe this code can be moved to sources/utils.c if #638 is merged before? Then we would have the harness function alone in this file.

Reply:
I'm open to moving this function depending on what everyone would prefer @feliperodri @alex-chew. The reason this function is in the harness for both this and the hdr_parse proofs is because the type of data structure generated by array_list_generator can change depending on the proof. So the general `aws_cryptosdk_hash_elems_array_init_stub.c `lives in the stub directory but the `array_list_generator` that generates the array elements lives in the harness itself. In the proofs I wrote, we're always generating elements of the enc_ctx which is why they look so similar. Even within the proofs I wrote, there is a small difference about whether the length is bounded by UINT8_MAX and UINT16_MAX. Nevertheless, so far we're really only used this array_list_item_generator in a similar way to the above, so we could move it to a `array_list_item_generator_enc_ctx_stub` or something of the kind to reuse it and make sure it's clearly documented.

Reply:
That makes sense to me. I agree this subtle distinction between generators needs to be well documented.